### ISSUE TITLE
Ship To State Issues

### ISSUE THREAD

**Issue by Host99** [taxcloud/ZenCart#3]
The problem we are having still is that ship to state is throwing crazy errors. "Error encountered looking up tax amount The Ship To State (Pe) is not valid" and other states as well.

All configurations match to the letter and it still throws the error. We have to disable the entire module for tax rates to work normally without TaxCloud.

If you plan to create a add on please keep it working, this not only makes TaxCloud look risky but will cause revenue issues. All files we have are your latest files. At this time we have recommended several customers not use this add on until you have fixed your issues.

Overall, it is a great add on, but it needs to be fully reworked to not fail as much.

Good Luck

**Comment by retched** [comment]
You get this error because PE is not a valid state code. Using an add-on like drop down state selector would work better here. Otherwise, Pennsylvania which should be PA will interpret as PE which is not valid. Also New Jersey would come up as Nebraska since NE is the first two letters.

**Comment by Host99** [comment]
As I mentioned clearly above ALL configurations match to the letter. We are not going to make a statement without thoroughly checking for any possible mess ups. The screenshot below clearly shows the setting we currently have in use.

![capture](https://cloud.githubusercontent.com/assets/6235393/5601204/7c910ee6-92c0-11e4-849b-5fd5678f164e.JPG)

**Comment by retched** [comment]
Was this module added to an already existing store with addresses already?

**Comment by Host99** [comment]
Yes, and  it worked perfectly fine until 1.5.3 and 1.54 upgrade. This mod has been in use for a year now. And the latest problem is a new customer using the drop down state options.

**Comment by retched** [comment]
Which means something in the 1.5.4 upgrade changed. I haven't done an upgrade yet so I guess I'll see when I take my site down in a few hours. But that's the only thing that I can think of. Something was changed which is causing the state full name instead of the state abbreviations from being sent out.

**Comment by Host99** [comment]
It started in 1.5.1 a while back disabled it for a couple days emailed someone and he called then reactivated then worked fine until today and now it does not work at all. Right now no taxes are being applied for New York orders at all.

**Comment by retched** [comment]
That's I'm using 1.5.3 with New York Taxes. And I didn't modify that many files, except for discounting.

**Comment by Host99** [comment]
no files that deal with taxes or discounting have been altered

**Comment by Host99** [comment]
Now its working again. Something is hay wired that is for sure

![capture](https://cloud.githubusercontent.com/assets/6235393/5601726/aa9f2d5a-92e2-11e4-8a01-ec9f432acf18.JPG)

**Comment by taxcloud** [comment]
Ship to state MUST be the two character abbreviation, NOT the complete
state name. TaxCloud will truncate past the first two characters. Please be
sure the ZenCart configuration is setup correctly (it is a configuration
option)

Go to your ZenCart admin - configuration - customers detail and set:

State - Always display as pulldown? = true.

Then, all atates will be sent to TaxCloud as two characters.

On Friday, January 2, 2015, Host99 <notifications@github.com
<javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:

> The problem we are having still is that ship to state is throwing crazy
> errors. "Error encountered looking up tax amount The Ship To State (Pe) is
> not valid" and other states as well.
>
> All configurations match to the letter and it still throws the error. We
> have to disable the entire module for tax rates to work normally without
> TaxCloud.
>
> If you plan to create a add on please keep it working, this not only makes
> TaxCloud look risky but will cause revenue issues. All files we have are
> your latest files. At this time we have recommended several customers not
> use this add on until you have fixed your issues.
>
> Overall, it is a great add on, but it needs to be fully reworked to not
> fail as much.
>
> Good Luck
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/taxcloud/ZenCart/issues/3>.
>


-- 
*R. David L. Campbell*
Chief Executive Officer
The Federal Tax Authority (FedTax <http://FedTax.net>)
Main Line: (206) 452-1686 (Ext. 1516)
Direct Dial: *(203) 883-1153*
Mobile: (206) 409-5866
--------------------------------------------------
TaxCloud <https://taxcloud.net>® is the internet's *first and only
completely free *and easy-to-use sales tax management service.
FedTax and TaxCloud are registered trademark of the Federal Tax Authority,
LLC.
Certain features of TaxCloud are protected by United States Patent No.
7,730,089.

CONFIDENTIALITY NOTICE: This message and any attachments may contain
confidential information protected by attorney-client or other contractual
privilege. If you believe that this message has been sent to you in error,
please reply to the sender that you received the message in error - Then
delete it. Thank you.

**Comment by Host99** [comment]
Yes, scroll above I posted the screenshot of that configuration.

**Comment by Host99** [comment]
I tried several customers different states it seems to do it with the state Pennsylvania more then most. That6 said, by going here https://www.pettreatsplus.com/index.php?main_page=login you will see that it is clearly the drop down list. Use a Pennsylvania address of some sort then try to check out it should show. Right now we will leave it disabled though this customer lost 10 orders already because of it.

**Comment by Host99** [comment]
Here is the error it displays

![capture](https://cloud.githubusercontent.com/assets/6235393/5601948/db5c093a-92f7-11e4-9a4f-ee91eb931ea6.JPG)

**Comment by Host99** [comment]
This may also help. Under My Personal Address Book it shows PA as the state but when you go to edit the address it shows the complete drop down list and not PA. Weird?

My Personal Address Book
![capture](https://cloud.githubusercontent.com/assets/6235393/5601956/3b56a61a-92f8-11e4-8564-29eec0e236a3.JPG)

Editing My Personal Address Book
![capture](https://cloud.githubusercontent.com/assets/6235393/5601958/50f4c5e2-92f8-11e4-9c6d-248e7e477839.JPG)

May be this will help? Some how.

**Comment by travvy** [comment]
Has this issue been resolved?  When customers us North Carolina, it is throwing the same error, stating that NO is not a valid state.  It is pulling NO from North when it should be NC.

**Comment by travvy** [comment]
Also note that this error is doing it for NY to.  It is pulling it up as NE.  I believe this is randomly happening and address dependent.  I made up a fake NY address but with a valid NY zip code, the error threw up.  NOw I went ahead and pulled a recent customer that made a NY order and mocked their address and it went through without a hitch.  So it doesn't sound like a ZC cart issue but the ZC module issue and maybe the tax data they are using.

**Comment by taxcloud** [comment]
Hi travvy and Host99,

**Thank you** for all of your input on this matter issue.

I think I know what is going on  - VerifyAddress is failing to validate the provided address, and then bad things unfold downstream...

1. Our **VerifyAddress API** actually accepts complete state names (New York, Pennsylvania, South Dakota, etc.), and it responds back with the correct two-character abbreviation (NY,PA,SD,etc.), which ZenCart is dutifully updates into its local variable for the 'ShipToState'.
2. However, **if VerifyAddress fails**, then the 'ShipToState' variable *never gets updated* to the correct two-character abbreviation.
3. Then, the process continues with a **Lookup API** call using the customer provided address (using the full state name, rather than the **required** two-character abbreviation).
4. Lookup API **rejects** request because it *automatically truncates* the state parameter to the first two characters. This unfortunate mechanism instantly turns:
   - :boom:  **New York** becomes **Nebraska**, and
   - :boom:  **Pennsylvania** becomes **Peru**, and
   - :boom:  **South Dakota** becomes **Somalia**!

We need to track down the VerifyAddress call and be sure to do **something better** (even if it's just a dumb string match/replace method to convert state names to state abbreviations) when the following **VerifiedAddress** response comes back
```
<VerifiedAddress>
<ErrNumber>97</ErrNumber>
<ErrDescription>Address can't be verified</ErrDescription>
</VerifiedAddress>
``` 
Thanks.
